### PR TITLE
Lab 3 sorting

### PR DESCRIPTION
**Pull request by don-tay**
## Changes in this PR

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change of an existing feature
- [ ] Refactor/Performance enhancements

## Overview

- Add handling of `order by` clause with `asc` and `desc`

### CODE REVIEW

File: src/simpledb/parse/Parser.java [hunk 1]
```diff
@@ -81,22 +81,29 @@ public Predicate predicate() {
...
       if (lex.matchKeyword("where")) {
          lex.eatKeyword("where");
          pred = predicate();
       }
-      return new QueryData(fields, tables, pred);
+      if (lex.matchKeyword("order")) {
+         lex.eatKeyword("order");
+         lex.eatKeyword("by");
+         // TODO: handle optional 'asc' and 'desc' keywords
+         sortfields = selectList();
```
Comment (don-tay) [inline src/simpledb/parse/Parser.java:102]: I foresee reimplementing this as a list of sortfields instead, to handle an optional `asc/desc` keyword.

For this, I intend to:
1. Create a `SortField` class comprising of the attributes `field` and `order`(default to `asc`)
2. Modify the above code to take a `List<SortField>` and a helper method to eat the keywords (similar to `selectList()`)
3. Modify `QueryData` class to hold the attribute `List<SortField> sortfields`

**Comment by don-tay** [comment]
Next steps:
1. Update `SortPlan` to take in a `List<SortField>` instead. 
2. Refactor dependencies `GroupByPlan`, `MergeJoinPlan` and `BasicQueryPlanner`
3. Modify `RecordComparator` to support `List<SortField>` and reverse order when `SortField` `order` attribute is `desc`

**Comment by don-tay** [comment]
Following queries work 
```
select SId, Sname, GradYear from student order by GradYear asc, Sname desc
select sid, sname from student order by sname desc, sid
select sid, sname from student order by sname desc, sid asc
```
Do re-run `CreateStudentDB` beforehand.

However, when running any order by query around 4-5 times, I am getting a `BufferAbortException`.

File: src/simpledb/plan/BasicQueryPlanner.java [hunk 2]
```diff
@@ -2,49 +2,54 @@
...
+      // Step 3: Add a selection plan for the predicate
       p = new SelectPlan(p, data.pred());
-      
-      //Step 4: Project on the field names
+
+      // Step 4: Project on the field names
       p = new ProjectPlan(p, data.fields());
+
+      // Step 5: Sort on the records
+      p = new SortPlan(tx, p, data.sortfields());
```
Comment (tishyakhanna97) [inline src/simpledb/plan/BasicQueryPlanner.java:51]: think this has to be optional
Screenshot 2022-02-13 at 2 30 51 AM
l

**Review by xinyee20 — APPROVED** [review]
lgtm!

**Comment by don-tay** [comment]
Another issue with this implementation is that it is unable to sort on fields which are not projected ie.
`SELECT sid FROM student ORDER BY sname` fails to run because sname is not in the `SELECT clause. 

This can be resolved by swapping the order of `ProjectPlan` and `SortPlan` in `BasicQueryPlanner`

**Comment by don-tay** [comment]
Buffer overflow bug has ben fixed in the above commit: https://github.com/don-tay/simpledb/pull/9/commits/3dae6971d6a53509860ecd9632974f406593ed9d

Additiionally, sorting now works on non-projected fields.

**Pull request merged**